sdk/eventgrid/eventgrid/src/consumer.ts

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Serializer } from "@azure/core-http";
+import { CloudEvent, EventGridEvent, CustomEventDataDecoder } from "./models";
+import {
+  EventGridEvent as EventGridEventMapper,
+  CloudEvent as CloudEventMapper
+} from "./generated/models/mappers";
+import { parseAndWrap } from "./util";
+import { systemDecoders } from "./systemEventDecoders";
+
+const serializer = new Serializer();
+
+function validateRequiredStringProperty(o: any, propertyName: string): void {
+  if (typeof o[propertyName] === "undefined") {
+    throw new TypeError(`event is missing required property '${propertyName}'`);
+  }
+
+  if (typeof o[propertyName] !== "string") {
+    throw new TypeError(
+      `event property '${propertyName} should be a 'string', but is '${typeof o[propertyName]}'`
+    );
+  }
+}
+
+function validateRequiredAnyProperty(o: any, propertyName: string): void {
+  if (typeof o[propertyName] === "undefined") {
+    throw new TypeError(`event is missing required property '${propertyName}'`);
+  }
+}
+
+function validateOptionalStringProperty(propertyName: string, o: any): void {
+  if (typeof o[propertyName] !== "undefined" && typeof o[propertyName] !== "string") {
+    throw new TypeError(
+      `event property '${propertyName}' should be a 'string' but it is a '${typeof o[
+        propertyName
+      ]}'`
+    );
+  }
+}
+
+/**
+ * TODO(matell): Document this.
+ */
+export class EventGridConsumer {
+  readonly customDecoders: Record<string, CustomEventDataDecoder>;
+  constructor(decoders: Record<string, CustomEventDataDecoder> = {}) {
+    this.customDecoders = decoders;
+  }
+
+  /**
+   * Decodes events encoded in the Event Grid schema.
+   *
+   * @param encodedEvents the JSON encoded representation of either a single event or an array of
+   * events, encoded in the Event Grid Schema.
+   */
+  public async decodeEventGridEvents(encodedEvents: string): Promise<EventGridEvent<unknown>[]>;
+  /**
+   * Decodes events encoded in the Event Grid schema.
+   *
+   * @param encodedEvents an object representing a single event, encoded in the Event Grid schema.
+   */
+  public async decodeEventGridEvents(encodedEvents: object): Promise<EventGridEvent<unknown>[]>;
+  public async decodeEventGridEvents(
+    encodedEvents: string | object
+  ): Promise<EventGridEvent<unknown>[]> {
+    const decodedArray = parseAndWrap(encodedEvents);
+
+    const events: EventGridEvent<unknown>[] = [];
+
+    for (const o of decodedArray) {
+      if (typeof o !== "object") {
+        throw new TypeError("event is not an object");
+      }
+
+      validateRequiredStringProperty(o, "eventType");
```

Comment:
perhaps it would be useful to have a syntax for checking multiple required string properties (e.g. with an array argument)